ssl/t1_lib.c

```diff
@@ -2639,30 +2660,130 @@ static int has_usable_cert(SSL *s, const SIGALG_LOOKUP *sig, int idx)
                 continue;
 
             ERR_set_mark();
-            supported = EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey,
-                                                     mdnid);
+            supported = EVP_PKEY_supports_digest_nid(pkey, mdnid);
+            ERR_pop_to_mark();
             if (supported == 0)
                 continue;
-            else if (supported < 0)
-            {
-                /* If it didn't report a mandatory NID, for whatever reasons,
-                 * just clear the error and allow all hashes to be used. */
-                ERR_pop_to_mark();
-            }
+            /*
+             * If it didn't report a mandatory NID (supported < 0), for
+             * whatever reasons, we just ignore the error and allow all
+             * hashes to be used.
+             */
             return 1;
         }
         return 0;
     }
-    supported = EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey,
-                                             sig->hash);
+    ERR_set_mark();
+    supported = EVP_PKEY_supports_digest_nid(pkey, sig->hash);
```

Comment:
(I think this check would be enough to do what David needed in #7408)

Reply:
OK, reproduced. I still need to put a breakpoint on `EVP_PKEY_supports_digest_nid()` and manually make it say 'no' to SHA512 because neither @jejb nor @andreasfuchsSIT have yet fixed their TPM2 engine to do that for itself. And OpenSSL doesn't seem to default to ECDSA+SHA512 any more, but I can force it:

```
openssl s_server -key ~/git/openconnect/tests/certs/server-key.pem -cert ~/git/openconnect/tests/certs/server-cert.pem -accept 8443 -sigalgs RSA+SHA256:ECDSA+SHA512:ECDSA+SHA256 -verify 1 -tls1_2
```
```
LD_LIBRARY_PATH=`pwd` gdb --args apps/openssl s_client -engine tpm2 -keyform engine -key ~/git/openssl_tpm2_engine/ec.tss -cert ~/git/openconnect/tests/certs/ec-cert.pem -connect localhost:8443
```
I can confirm what @kaduk says, that the check at line 2677 is the interesting one. The other one is bogus.

How does this look?

https://github.com/dwmw2/openssl/commit/7403d434fe37429c19e7fd3675be03cdee757bed

Reply:
I submitted that as #9672. I have a backport for 1.1.1 too, which I'll submit if this passes muster.